test/libsolidity/semanticTests/various/selfdestruct_pre_cancun_redeploy.sol

```diff
@@ -0,0 +1,161 @@
+contract Factory {
+    event Deployed(address, bytes32);
+
+    function deploy(bytes32 _salt) external payable returns (address implAddr) {
+        // NOTE: The bytecode of contract C is used here instead of `type(C).creationCode` since the address calculation depends on the precise init code
+        // and that will change in our test framework between legacy and via-IR codegen and via optimized vs non-optimized.
+        //contract C {
+        //    constructor() payable {}
+        //    receive() external payable {}
+        //    function terminate() external {
+        //        selfdestruct(payable(msg.sender));
+        //    }
+        //}
+        bytes memory initCode =
+            hex"608080604052606c908160108239f3fe60043610156013575b36156011575f80"
+            hex"fd5b005b5f3560e01c630c08bf8803600857346032575f366003190112603257"
+            hex"33ff5b5f80fdfea26469706673582212206e19160634438b12a36d2fa77cda05"
+            hex"0e2c46846d03e41b19aab5a6207d18356564736f6c63430008180033";
+
+        address target = address(uint160(uint256(keccak256(abi.encodePacked(
+            bytes1(0xff),
+            address(this),
+            _salt,
+            keccak256(abi.encodePacked(initCode))
+        )))));
+
+        assembly {
+            implAddr := create2(callvalue(), add(initCode, 0x20), mload(initCode), _salt)
+            if iszero(extcodesize(implAddr)) {
+                revert(0, 0)
+            }
+        }
+        assert(address(implAddr) == target);
+        emit Deployed(implAddr, _salt);
+    }
+}
+
+interface IC {
+    function terminate() external;
+}
+
+contract D {
+    enum CallKind {
+        CREATE2,
+        CALL,
+        DELEGATECALL
+    }
+
+    Factory public factory;
+    IC public c;
+    bytes32 constant salt = hex"1234";
+
+    constructor() payable {
+        factory = new Factory();
+    }
+
+    function deploy_create2() public payable {
+        // NOTE: `create2` cannot be used anymore to redeploy a contract in the same place in order to make it upgradable.
+        // Thus, an attempt to redeploy using `create2` must revert for EVM >= cancun.
+        c = IC(factory.deploy{value: 1 ether}(salt));
+    }
+
+    function deploy_call() public payable returns (bool) {
```

Comment:
> Also it seems that the difference is not really in whether the contract gets redeployed. It's also not in `create2`, because the first test has one case like this too. It's in he use of specifically low-level `create2`. Test naming should reflect that.

What do you mean by that? 

> Any reason to include the `deploy_call()` and `deploy_delegatecall()` variants here but not in the post-cancun test? I expected them to be identical but they are not. Actually, the other pair of tests also has slight differences (though in their case mostly in comment wording).

Those would not work in the same test since the address would be the same, and the contract would still exist after `selfdestruct`. The deploy would always fail. I believe this is why they mention in the EIP that redeploy with `create2` would not work anymore when performed in different transactions. This is why the `post-cancun` test for redeploy only checks one case and not the low-level calls case. But I can add separate tests for low-level calls if you think it is necessary, but I'm considering removing it based on your last comment and just leaving one redeploy case.

Reply:
> What do you mean by that? 

I mean that the other test also has a test that redeploys under the same address (`new` with salt) so both the fact or redeploying is not what differentiates these tests as the name implies.

> The deploy would always fail.

You can still have an expectation that it fails.

But yeah, I'm not saying you should add more :) I don't think so many low-level call tests are necessary here :)

Reply:
> I mean that the other file also has a test that redeploys under the same address (using `new` with salt) so the fact of redeploying is not what differentiates these tests as the name implies.

But there, the contract get deleted, and here it doesn't when it is not in the same transaction.

Reply:
I don't understand. Both cases delete the test with `terminate()` before recreating it.

Reply:
Yes, but they call `deploy_create2` right after, which should succeed for `pre_cancun` but fail for `post_cancun`. What better name would you suggest for them?

Reply:
I'd suggest to include that `deploy_create2()` after `terminate()` in the original tests. I didn't even realize that this was the relevant difference. The fact that you're using low-level `create2()` (which includes a pretty nasty setup with hard-coded bytecode) seemed much more relevant.

When you do that the difference between tests is really only in whether `create2` is low or high level. And like I said before, not sure if doing it the low-level way actually gives us anything here. For me it just makes tests more complicated to review :)

Reply:
Yeah, I just used the low-level `create2` because without it our test suite cannot generate the same contract address, so I couldn't effectively redeploy the contract at same address.